worker/raft/apply.go

```diff
@@ -0,0 +1,125 @@
+	// for the command to be started. This must be run on the leader or it
+	// will fail.
+	Apply([]byte, time.Duration) raft.ApplyFuture
+}
+
+// Applier applies a new operation against a raft instance.
+type Applier struct {
+	raft         Raft
+	notifyTarget raftlease.NotifyTarget
+	clock        clock.Clock
+	logger       Logger
+}
+
+// NewApplier creates a new Applier.
+func NewApplier(raft Raft, target raftlease.NotifyTarget, clock clock.Clock, logger Logger) LeaseApplier {
+	return &Applier{
+		raft:         raft,
+		notifyTarget: target,
+		clock:        clock,
+		logger:       logger,
+	}
+}
+
+// ApplyOperation applies an lease opeartion against the raft instance. If the
+// raft instance isn't the leader, then an error is returned with the leader
+// information if available.
+// This Raft spec outlines this "The first option, which we recommend ..., is
+// for the server to reject the request and return to the client the address of
+// the leader, if known." (see 6.2.1).
+// If the leader is the current raft instance, then attempt to apply it to
+// the fsm.
+func (a *Applier) ApplyOperation(op queue.Operation) error {
+	if state := a.raft.State(); state != raft.Leader {
+		leaderAddress := a.raft.Leader()
+
+		a.logger.Infof("Attempt to apply the lease failed, we're not the leader. State: %v, Leader: %v", state, leaderAddress)
+
+		// If the leaderAddress is empty, this implies that either we don't
+		// have a leader or there is no raft cluster setup.
+		if leaderAddress == "" {
+			// Return back we don't have a leader and then it's up to the client
+			// to work out what to do next.
+			return NewNotLeaderError("", "")
+		}
+
+		// If we have a leader address, we hope that we can use that leader
+		// address to locate the associate server ID. The server ID can be used
+		// as a mapping for the machine ID.
+		future := a.raft.GetConfiguration()
+		if err := future.Error(); err != nil {
+			return errors.Trace(err)
+		}
+
+		config := future.Configuration()
+
+		// If no leader ID is located that could imply that the leader has gone
+		// away during the raft.State call and the GetConfiguration call. If
+		// this is the case, return the leader address and no leader ID and get
+		// the client to figure out the best approach.
+		var leaderID string
+		for _, server := range config.Servers {
+			if server.Address == leaderAddress {
+				leaderID = string(server.ID)
+				break
+			}
+		}
+
+		return NewNotLeaderError(string(leaderAddress), leaderID)
+	}
+
+	a.logger.Tracef("Applying command %v", string(op.Command))
+
+	timeout := op.Deadline.Sub(a.clock.Now())
+	future := a.raft.Apply(op.Command, timeout)
+	if err := future.Error(); err != nil {
+		return errors.Trace(err)
+	}
+
+	response := future.Response()
+	fsmResponse, ok := response.(raftlease.FSMResponse)
+	if !ok {
+		// This should never happen.
+		panic(errors.Errorf("programming error: expected an FSMResponse, got %T: %#v", response, response))
```

Comment:
Probably rather log it at critical level.

Reply:
I don't mind, this shouldn't happen.